Sources/Fullscreen/FullscreenGallery/FullscreenGalleryOverlayView.swift

```diff
@@ -0,0 +1,231 @@
+    // during the dismissal animation.
+    private lazy var previewViewHiddenConstraint = previewView.topAnchor.constraint(equalTo: bottomAnchor, constant: .mediumLargeSpacing)
+
+    // MARK: - Init
+
+    override init(frame: CGRect) {
+        fatalError("not implemented")
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("not implemented")
+    }
+
+    init(withPreviewViewVisible previewViewVisible: Bool) {
+        self.previewViewInitiallyVisible = previewViewVisible
+        self.previewViewVisible = previewViewVisible
+        super.init(frame: .zero)
+        setup()
+    }
+
+    private func setup() {
+        addSubview(captionLabel)
+        addSubview(dismissButton)
+        addSubview(previewView)
+
+        let initialPreviewViewVisibilityConstraint = previewViewVisible
+            ? previewViewVisibleConstraint
+            : previewViewHiddenConstraint
+
+        NSLayoutConstraint.activate([
+            captionLabel.centerXAnchor.constraint(equalTo: safeLayoutGuide.centerXAnchor),
+            captionLabel.widthAnchor.constraint(lessThanOrEqualTo: safeLayoutGuide.widthAnchor, constant: -(2 * CGFloat.mediumLargeSpacing)),
+            captionLabel.bottomAnchor.constraint(lessThanOrEqualTo: safeLayoutGuide.bottomAnchor, constant: -.mediumSpacing),
+            captionLabel.bottomAnchor.constraint(lessThanOrEqualTo: previewView.topAnchor, constant: -.mediumSpacing),
+
+            dismissButton.leadingAnchor.constraint(equalTo: safeLayoutGuide.leadingAnchor, constant: .mediumLargeSpacing),
+            dismissButton.topAnchor.constraint(equalTo: safeLayoutGuide.topAnchor, constant: .mediumLargeSpacing),
+            dismissButton.widthAnchor.constraint(equalToConstant: FullscreenGalleryOverlayView.dismissButtonSize),
+            dismissButton.heightAnchor.constraint(equalToConstant: FullscreenGalleryOverlayView.dismissButtonSize),
+
+            previewView.leadingAnchor.constraint(lessThanOrEqualTo: leadingAnchor),
+            previewView.trailingAnchor.constraint(lessThanOrEqualTo: trailingAnchor),
+            previewView.centerXAnchor.constraint(equalTo: centerXAnchor),
+            initialPreviewViewVisibilityConstraint
+        ])
+    }
+
+    // MARK: - Public methods
+
+    func scrollToImage(atIndex index: Int, animated: Bool) {
+        previewView.scrollToItem(atIndex: index, animated: animated)
+        setCaptionLabel(index: index)
+    }
+
+    func setThumbnailPreviewsVisible(_ visible: Bool) {
+        guard visible != previewViewVisible else {
+            return
+        }
+
+        if visible {
+            NSLayoutConstraint.deactivate([previewViewHiddenConstraint])
+            NSLayoutConstraint.activate([previewViewVisibleConstraint])
+        } else {
+            NSLayoutConstraint.deactivate([previewViewVisibleConstraint])
+            NSLayoutConstraint.activate([previewViewHiddenConstraint])
+        }
```

Comment:
This could be rewritten as

```swift
previewViewVisibleConstraint.isActive = visible
previewViewHiddenConstraint.isActive = !visible
```

Reply:
I didn't know that, thanks! This makes the constraints invalid when showing the thumbnails however, as the `hidden` and `visible` constraints are active simultaneously. This makes auto-layout complain and attempt to break one constraint to resolve the conflict - it seems to always break the correct constraint, but I don't think that's a die we want to roll in production... 🙂